ctapipe/calib/camera/r1.py

```diff
@@ -313,65 +320,7 @@ def real_calibrate(self, event):
             event.r1.tel[self.telid].waveform = self._r1_wf
 
 
-class CameraR1CalibratorFactory(Factory):
-    """
-    The R1 calibrator `ctapipe.core.factory.Factory`. This
-    `ctapipe.core.factory.Factory` allows the correct
-    `CameraR1Calibrator` to be obtained for the data investigated.
-
-    Additional filepaths are required by some cameras for R1 calibration. Due
-    to the current inplementation of `ctapipe.core.factory.Factory`, every
-    trait that could
-    possibly be required for a child `ctapipe.core.component.Component` of
-    `CameraR1Calibrator` is
-    included in this `ctapipe.core.factory.Factory`. The
-    `CameraR1Calibrator` specific to a
-    camera type should then define how/if that filepath should be used. The
-    format of the file is not restricted, and the file can be read from inside
-    ctapipe, or can call a different library created by the camera teams for
-    the calibration of their camera.
-    """
-    base = CameraR1Calibrator
-    custom_product_help = ('R1 Calibrator to use. If None then a '
-                           'calibrator will either be selected based on the '
-                           'supplied EventSource, or will default to '
-                           '"NullR1Calibrator".')
-
-    def __init__(self, config=None, tool=None, eventsource=None, **kwargs):
-        """
-        Parameters
-        ----------
-        config : traitlets.loader.Config
-            Configuration specified by config file or cmdline arguments.
-            Used to set traitlet values.
-            Set to None if no configuration to pass.
-        tool : ctapipe.core.Tool
-            Tool executable that is calling this component.
-            Passes the correct logger to the component.
-            Set to None if no Tool to pass.
-        eventsource : ctapipe.io.eventsource.EventSource
-            EventSource that is being used to read the events. The EventSource
-            contains information (such as metadata or inst) which indicates
-            the appropriate R1Calibrator to use.
-        kwargs
-
-        """
-
-        super().__init__(config, tool, **kwargs)
-        if eventsource and not issubclass(type(eventsource), EventSource):
-            raise TypeError(
-                "eventsource must be a ctapipe.io.eventsource.EventSource"
-            )
-        self.eventsource = eventsource
-
-    def _get_product_name(self):
-        try:
-            return super()._get_product_name()
-        except AttributeError:
-            es = self.eventsource
-            if es:
-                if es.metadata['is_simulation']:
-                    return 'HESSIOR1Calibrator'
-                elif es.__class__.__name__ == "TargetIOEventSource":
-                    return 'TargetIOR1Calibrator'
-            return 'NullR1Calibrator'
+__all__ = [
+    cls.__name__
+    for cls in non_abstract_children(CameraR1Calibrator)
+]
```

Comment:
is it possible to put this boilerplate code in a metaclass as we did before? Then, it's transparent to a user, and we don't have to remember to cut and paste this in.

Reply:
What I mean by that is a new developer should not have to paste that loop into each class that needs it, they should just set a metaclass to "Producible" or something like that, and the metaclass automatically sets __all__

Reply:
I am sorry there seems to be a misunderstanding here, the in the master, this part looks like this:
https://github.com/cta-observatory/ctapipe/blob/c5f609b8bf31d5fe8d17577b020949d40ffe44b2/ctapipe/calib/camera/r1.py#L22-L25

The advantage of having this loop is only, that a developer of a new CameraR1Calibrator does not need to remember to put the name of the new class into `__all__`, but we can of course revert back to what is in the master.

> a new developer should not have to paste that loop into each class that needs it

Absolutely. And they do not ... they can do whatever they want. 

I see putting this loop into `__all__` is confusing for reviewers and it is unrelated to the actual PR. I will remove it.

Reply:
This is fixed now